tunnel/tunnel.go

```diff
@@ -50,8 +50,9 @@ type regSafeStream struct {
 // Send blocks until it sends the provided data to the stream.
 func (r *regSafeStream) Send(data *tpb.RegisterOp) error {
 	r.w.Lock()
-	defer r.w.Unlock()
-	return r.regStream.Send(data)
+	err := r.regStream.Send(data)
+	r.w.Unlock()
```

Comment:
not sure why move Unlock() from defer. `defer` ensures proper cleanup even the `send` operation fails.

Reply:
similarly for those below